### DISCUSSION TITLE
[SOLVED] Database backwards compatibility / copy URLs of articles in 4.0.4 (Raspberry Pi version)

### DISCUSSION THREAD

**Discussion by Bro-Account** [martinrotter/rssguard#892] [Q&A] [answered]
Hi,

First I have to say that I'm so glad I found this application. It looks like I finally have a replacement for the quite outdated RSSOwl(nix), with a major new benefit: Raspberry Pi compatibility. I want to run it on my RPi, so I don't need to keep my PC running to not miss any new items. (*1)

Due to the limited resources on the RPi, I'd prefer to just copy the link list to the clipboard and open it on my Desktop browser instead. Unfortunately, the RPi version is still on 4.0.4 and it seems it doesn't offer this feature yet.

So I thought, as a workaround, before I go through the feeds I might export the DB on the RPi, import it on my Desktop on the newer version and use it there. When I'm done, I'd export the changes again to the RPi instance. In my test using a Windows 4.3.2 instance and importing its backup DB to a Windows 4.0.4 instance this didn't work, 4.0.4 didn't respond anymore after importing the DB and restarting.

So I assume the DB isn't backwards compatible (even though there was no warning)?

Is there any other way to do this until the RPi gets a newer version? Right now I can only think of exporting the current DB to Windows for copying the items, but then making all changes to the same selected items on the RPi.

Thanks a lot!

*1)
Unfortunately there are sites with lots of content, but a very short number of recent items in the feed. Some of those completely refill the feed in 10-20 minutes at peak times.

**Comment by martinrotter** [comment]
Hi.

Well. RSS Guard database is FORWARD compatible of course, As for backwards compatibility - database schema is sometimes updated to accomodate new features - new columns are sometimes added, some tables are added etc.

Here are current scripts which are responsible for updating DB tables to "current" schema version.

https://github.com/martinrotter/rssguard/tree/master/resources/sql

Sadly if in some version some columns are added, then naturally old versions of RSS Guard will not be able to read that new DB.

To my knowledge no widely used applications ensure backwards compatibility. There might be some compatility for some previous versions but 4.0.4 is just tooooo old.

My suggestion would be to recompile RSS Guard for your system on Raspberry pi by hand? Compilation does not really take that long, even on slow configurations.

**Comment by Bro-Account** [comment, marked as answer]
Thanks! I didn't really expect backwards compatibility. But it might be a good idea to add a DB version check to the Import/Export GUI. 

I think I have never compiled something on my own before. After dozens of iterations of error messages/web search/package installations I think I'm almost there, but not quite.

I was able to run the following commands with no errors:
rssguard-4.3.2$ cmake -DCMAKE_BUILD_TYPE="Release" .
rssguard-4.3.2$ cmake --build .

But at the final step I get

rssguard-4.3.2$ cmake --install .
-- Install configuration: "Release"
CMake Error at src/librssguard/cmake_install.cmake:46 (file):
  file INSTALL cannot find
  "[...]/rssguard-4.3.2/src/librssguard/CMakeFiles/CMakeRelink.dir/librssguard.so":
  No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:48 (include)

Any idea what I've missed? It might have something to do with this, as I don't really understand this, neither I can see the folders anywhere:

> Usage (out of source build type, we have two side by side folders:
>  empty "build-dir" and RSS Guard repository "rssguard-dir"):
>    b) RELEASE build for production use.
>      cd build-dir

**Reply by Bro-Account** [reply]
Finally figured out how this was meant. For anyone else as inexperienced as I am:

Create a new folder. Within the new folder, you will have to extract the contents of the source archive file. Right now with the current version this will extract the contents of the archive to a folder named `rssguard-4.3.2`. This is the rssguard-dir. Next to this folder, you will have to create another folder called `build-dir`.

Now you open the Terminal within the build-dir folder and run the following commands:

```
cmake -DCMAKE_BUILD_TYPE="Release" ../rssguard-4.3.2/
cmake --build .
```

This time the error message was different when running `cmake --install .`

```
bro@raspberrybro:~/Documents/rssguard/build-dir$ cmake --install .
-- Install configuration: "Release"
-- Installing: /usr/local/lib/librssguard.so
CMake Error at src/librssguard/cmake_install.cmake:52 (file):
  file INSTALL cannot copy file
  "/home/bro/Documents/rssguard/build-dir/src/librssguard/librssguard.so"
  to "/usr/local/lib/librssguard.so": Permission denied.
Call Stack (most recent call first):
  cmake_install.cmake:48 (include)
```


Tried again with `sudo cmake --install .`

and it worked this time. :)


When trying to run the installed application, nothing happened. Trying to run it from terminal revealed:

```
/usr/local/bin$ ./rssguard 
./rssguard: error while loading shared libraries: librssguard.so: cannot open shared object file: No such file or directory

```

Located the application in the build dir `~/Documents/rssguard/build-dir/src/rssguard$` and opened it from there instead and it worked finally. ❤️

**Comment by Bro-Account** [comment]
Unfortunately doing it the same way on the Raspberry Pi OS doesn't work. Does anyone have any idea what's the problem here?

```
Scanning dependencies of target app
[100%] Building CXX object src/rssguard/CMakeFiles/app.dir/app_autogen/mocs_compilation.cpp.o
[100%] Building CXX object src/rssguard/CMakeFiles/app.dir/main.cpp.o
[100%] Linking CXX executable rssguard
[100%] Built target app
bro@raspberrypi400:~/Documents/rssguard/build-dir $ sudo cmake --install .
-- Install configuration: "Release"
-- Installing: /usr/local/lib/aarch64-linux-gnu/librssguard.so
-- Installing: /usr/local/include/librssguard/services/abstract
-- Installing: /usr/local/include/librssguard/services/abstract/rootitem.h
-- Installing: /usr/local/include/librssguard/services/abstract/serviceroot.h
-- Installing: /usr/local/include/librssguard/services/abstract/serviceentrypoint.h
-- Installing: /usr/local/include/librssguard/services/abstract/category.h
-- Installing: /usr/local/include/librssguard/services/abstract/unreadnode.h
-- Installing: /usr/local/include/librssguard/services/abstract/labelsnode.h
-- Installing: /usr/local/include/librssguard/services/abstract/cacheforserviceroot.h
-- Installing: /usr/local/include/librssguard/services/abstract/recyclebin.h
-- Installing: /usr/local/include/librssguard/services/abstract/accountcheckmodel.h
-- Installing: /usr/local/include/librssguard/services/abstract/importantnode.h
-- Installing: /usr/local/include/librssguard/services/abstract/gui
-- Installing: /usr/local/include/librssguard/services/abstract/gui/custommessagepreviewer.h
-- Installing: /usr/local/include/librssguard/services/abstract/gui/formcategorydetails.h
-- Installing: /usr/local/include/librssguard/services/abstract/gui/formfeeddetails.h
-- Installing: /usr/local/include/librssguard/services/abstract/gui/authenticationdetails.h
-- Installing: /usr/local/include/librssguard/services/abstract/gui/formaccountdetails.h
-- Installing: /usr/local/include/librssguard/services/abstract/feed.h
-- Installing: /usr/local/include/librssguard/services/abstract/label.h
-- Installing: /usr/local/bin/rssguard
-- Set runtime path of "/usr/local/bin/rssguard" to ""
-- Installing: /usr/local/share/applications/io.github.martinrotter.rssguard.desktop
-- Installing: /usr/local/share/metainfo/io.github.martinrotter.rssguard.metainfo.xml
-- Installing: /usr/local/share/icons/hicolor/256x256/apps/io.github.martinrotter.rssguard.png
bro@raspberrypi400:~/Documents/rssguard/build-dir $ /usr/local/bin/rssguard
/usr/local/bin/rssguard: error while loading shared libraries: librssguard.so: cannot open shared object file: No such file or directory
bro@raspberrypi400:~/Documents/rssguard/build-dir $ cd src/rssguard/
bro@raspberrypi400:~/Documents/rssguard/build-dir/src/rssguard $ rssguard 
rssguard: error while loading shared libraries: librssguard.so: cannot open shared object file: No such file or directory
bro@raspberrypi400:~/Documents/rssguard/build-dir/src/rssguard $ 

```

**Reply by Bro-Account** [reply]
Nevermind, it's working now after rebooting my Raspberry Pi.
